### DISCUSSION TITLE
Binding refresh background task crashes if I try to use app.storage.user in binding backward function

### DISCUSSION THREAD

**Discussion by JohanAR** [zauberzeug/nicegui#3093] [Q&A] [answered]
### Question

I have some UI things which I want to have user-configurable, so I would like check app.storage.user whenever the value is updated and possibly invert it. This appears to crash a background task, but otherwise works as expected. Minimal reproduction project:

```py
import typing

from nicegui import app, binding, ui


class MyElement(ui.element):
    a_value = binding.BindableProperty(
        on_change=lambda sender, value: typing.cast(typing.Self, sender)._handle_value_change(value))

    def __init__(self):
        super().__init__()
        binding.bind_from(self, 'a_value',
                          app.storage.general, 'a_value',
                          backward=lambda v: app.storage.user.get('do_invert', False) == v)

    def _handle_value_change(self, new_value: bool) -> None:
        self.classes(replace=('bg-yellow' if new_value else 'bg-red'))


@ui.page('/')
def index():
    with MyElement():
        ui.switch('The value').bind_value(app.storage.general, 'a_value')
        ui.switch('Invert it').bind_value(app.storage.user, 'do_invert')


ui.run(storage_secret='seeecret')
```

**Comment by rodja** [comment, marked as answer]
Yes, it seems that in backward/forward you do not have access to the app.storage. This might be related to #2520. A workaround may be archived like this:

```py
class MyElement(ui.element):

    def __init__(self):
        super().__init__()
        self.a_value = False
        binding.bind_from(self, 'a_value', app.storage.general, 'a_value', backward=self.transform)

    def transform(self, new_value: bool) -> bool:
        self.classes(replace=('bg-yellow' if new_value else 'bg-red'))
        return new_value


@ui.page('/')
def index():
    with MyElement().classes('w-[400px]'):
        ui.switch('The value').bind_value(app.storage.general, 'a_value').bind_value(app.storage.user, 'do_invert')
        ui.switch('Invert it', value=False).bind_value_from(app.storage.user, 'do_invert', backward=lambda x: not x)
```

**Reply by JohanAR** [reply]
Thanks! Not sure if it would be technically complicated to allow storage access in forward/backward, but it would be nice with a clearer error message if that's easier to implement.

**Reply by falkoschindler** [reply]
I guess you've seen the error message "RuntimeError: app.storage.user needs a storage_secret passed in ui.run()" with your code. That's what we hope to improve with #2520 in version 2.0.0.

**Reply by JohanAR** [reply]
Yea, I got that error message. In retrospect I also think it's potentially confusing that the program continues to run after one of its background tasks crash, and I suspect that it might have some subtle impact on my app which seemingly still worked normally.
